### DISCUSSION TITLE
Editor Support

### DISCUSSION THREAD

**Discussion by ScottAwesome** [markdoc/markdoc#5] [Ideas]
Great work here by the Stripe team! Really like this as a documentation generating tool. 

My question (and idea) is I am wondering if there is any plans to support a VS Code and IntelliJ extension for the syntax? Since its "non-standard" markdown syntax and doesn't use any the pre-existing syntax of say, Liquid or Handlebars, the developer experience isn't as robust as something like say, MDX.

**Comment by segphault** [comment]
Yes, we built a Visual Studio Code extension for our internal use at Stripe that supports Markdoc syntax highlighting, validation, AST visualization, and a range of other features. It is currently a Stripe-specific solution, but we are planning to refactor it this year to make an open source Markdoc language server protocol implementation for the community, with support for VSCode and other LSP-compatible editors.

**Reply by christianscheuer** [reply]
That's awesome news! It would be great if you when doing the refactor kept in mind the use case where Markdoc is being edited via the Monaco editor in a browser context as part of a (custom) CMS, and not just VS Code on the desktop.

Perhaps we could leverage something like https://github.com/TypeFox/monaco-languageclient to connect to the language server (would require that it's Javascript and can run in the browser).

**Reply by addisonj** [reply]
@segphault Thanks so much for your work on markdoc.

I have started looking at building my own tooling for vscode to support some of these use cases and then came across this thread. 

I was curious if this is still planned? Or if you might have more details to share on what you have built?

I am doing some PoC work to help validate using Markdoc with a team of technical writers and being able to show what that editing experience could look like is a big piece.

Thanks again for this great project, looking forward to using it more :)

**Reply by rpaul-stripe** [reply]
This is still planned, but we don't have a timeline that we can share at this time. I'm happy to answer specific questions about the feature set if that would be helpful.

One of the challenges with providing a good tooling experience around Markdoc is that a lot of the functionality relies on the schema, which is arbitrary user-supplied code that can potentially have complicated dependencies. For our internal tooling, we generate a static build of our schema and bundle it directly into the extension, which is obviously not a viable solution for other users.

What we'd like to do instead is create a library that speaks the VSCode language server protocol and make it so that you can embed it directly in your own application, passing in your schema as a parameter. This would make it so that when you run your Markdoc site locally for preview purposes it also starts up a language service that the VSCode extension can attach to so you get validation etc with your own schema.

**Reply by romanr** [reply]
What could be a reason for not releasing it? It's not like there's abundance of tools supporting it, none so far after more than a year.    
Why release Markdoc to public at all if you are deliberately withholding the toolset that would help the adoption?

**Reply by mfix-stripe** [reply]
@romanr, @rpaul-stripe shared reasons above for why we aren't releasing our internal solution—it is not a viable one for Markdoc users outside of Stripe. Good news is, we are working on much better editor support that anyone can use, and we are hoping to release it it sometime in middle of this year.

**Reply by rpaul-stripe** [reply]
@romanr the official language server and VSCode extension are now available: https://github.com/markdoc/markdoc/discussions/407

The internal VSCode extension that we used at Stripe included our own schema as a hard-coded dependency and also hard-coded many assumptions about our routing system, metadata, and other characteristics of our publishing pipeline. It was simply not suitable for distribution to the outside world and would not have been useful to anyone else. The new version that we're releasing today is a complete rewrite that is designed to dynamically load an external Markdoc schema. It's significantly better and more complete than the tools that we've been using internally up to this point. I can appreciate frustration with the long wait for us to release this, but I can assure you that we were not arbitrarily withholding anything that would have been useful.

**Comment by mohitsinghs** [comment]
For anyone else looking for syntax highlighting in VSCode, I've released [one extension](https://marketplace.visualstudio.com/items?itemName=mohitsingh.markdoc) till official one is out. It's not great and have bugs but highlights are there for most of the things. The code lives [here](https://github.com/mohitsinghs/vscode-markdoc).

**Reply by mfix-stripe** [reply]
Looks nice @mohitsinghs! Cool VSCode theme, too. 💫

**Reply by mohitsinghs** [reply]
Thanks, Markdoc was exactly what I was looking for, and that theme is tokyonight.

**Reply by mohitsinghs** [reply]
So, I tried putting together a [language server](https://github.com/mohitsinghs/markdoc-ls) for the extension for fun. It relies on [my markdoc fork](https://github.com/mohitsinghs/markdoc) published on [github registry](https://github.com/mohitsinghs/markdoc/pkgs/npm/markdoc) It's a POC but I got validations working along with config.

---
![diagnostics](https://user-images.githubusercontent.com/4941333/191670516-697926b4-a182-4d8d-812c-00ee8629088b.png)

**Reply by nkohari-stripe** [reply]
Awesome work!

**Reply by mohitsinghs** [reply]
Thanks, I got formatter and part of completions working as well.

**Reply by mohitsinghs** [reply]
If anyone is missing it in neovim like me, I created [tree-sitter-markdoc](https://github.com/markdoc-extra/tree-sitter-markdoc).

![tree-sitter](https://user-images.githubusercontent.com/4941333/222037580-0cdc6350-1362-42ba-b359-b346d3ba877f.png)

**Reply by mfix-stripe** [reply]
This deserves it's own show and tell post in the Discussions forum 🙂

**Reply by mohitsinghs** [reply]
Thanks @mfix-stripe , started a [discussion](https://github.com/markdoc/markdoc/discussions/337) for this.

**Reply by louiss0** [reply]
> This deserves it's own show and tell post in the Discussions forum 🙂



> For anyone else looking for syntax highlighting in VSCode, I've released [one extension](https://marketplace.visualstudio.com/items?itemName=mohitsingh.markdoc) till official one is out. It's not great and have bugs but highlights are there for most of the things. The code lives [here](https://github.com/mohitsinghs/vscode-markdoc).

When you finish refactoring this extension can you please tell [Ben Holmes](https://twitter.com/BHolmesDev) from the [Astro](https://astro.build/) Team or the Next.js people. They both need this extension because Astro is going to support Markdoc soon.

**Comment by louiss0** [comment]
When you finish refactoring this extension can you please tell [Ben Holmes](https://twitter.com/BHolmesDev) from the Astro Team or the Next.js people. They both need this extension because Astro is going to support Markdoc soon.

**Comment by rpaul-stripe** [comment]
The official Visual Studio Code extension and language server are now available: https://github.com/markdoc/markdoc/discussions/407
